교수자 강의 생성 / 학생 수강 등록 기능 개발

1seyoung opened the pull request:
### 구현 내용
강의 생성 및 수강 신청 기능을 구현하였습니다. 교수자와 학생의 역할에 따른 API를 분리하고, 각각의 대시보드를 구현하였습니다.

📍강의 도메인 구현
Lecture 모델 구현 - UUID 기반의 고유 코드 생성
LectureStudent를 통한 수강생 관리

📍컨트롤러 분리

LectureController -> InstructorLectureController, StudentLectureController로 분리
공통 기능은 CommonLectureController에서 관리 (고민사항)

API 엔드포인트

GET /api/instructor/lectures: 교수자의 강의 목록 조회
POST /api/instructor/lectures: 새 강의 생성
GET /api/student/lectures: 학생의 수강 중인 강의 목록 조회
POST /api/student/lectures/join: 강의 참여 (수강신청)
GET /api/lectures/{identifier}: 강의 상세 정보 조회 (공통)

### 고민 포인트, 리뷰 시 참고 사항(코드)
이전 PR 리뷰 반영했습니다. 브랜치 관리 실수로 다시 올립니다. 

**주요 변경 사항**
src/main/java/com/gatheria/mapper/LectureMapper.java
```
  @Select("SELECT l.*, i.affiliation, m.name as instructor_name " +
      "FROM lectures l " +
      "JOIN lecture_students ls ON l.id = ls.lecture_id " +
      "JOIN instructors i ON l.instructor_id = i.id " +
      "JOIN members m ON i.member_id = m.id " +
      "WHERE ls.student_id = #{studentId}")
  List<Lecture> findByStudentId(@Param("studentId") Long studentId);
```

> 가능한 조인보다는 단건 조회를 여러 번 활용하여 데이터를 가져오시면 좋겠습니다.
> 자바 코드 단에서 member 가져오고, instructor 가져오고 그것을 기반으로 lecture 가져오고 이렇게 한다고 가정하면 쿼리가 3번 발생하니 비효율적으로 보일 수 있으나, 어차피 인덱싱이 동작하므로 크게 문제가 없습니다.
> 그보다는 쿼리를 단순하게 가져가야 유지 보수성도 좋고, 재사용성도 좋기때문에 이를 추천합니다!

-> 단건 조회 이용하는 것으로 코드 수정 완료



### 메인 리뷰어 지정 및 Due Date
@blue000927 

### 관련 이슈
#9 
#5 수업
#10 수업등록

1seyoung committed:
feat : JwtUtil 기능 구현
- token 정보에 memberid 추가
- 토큰 검증 메소드 구현
- 토큰에서 정보 parsing (extractEmail, extractRole, extractMemberId)

1seyoung committed:
feat : 수업 생성 컨트롤러 구현

1seyoung committed:
chore : jwt + security 문제 발생

1seyoung committed:
merge: resolve conflicts from develop into feature/lecture

1seyoung committed:
chore : gitignore 추가

1seyoung committed:
chore : gitignore 추가

1seyoung committed:
chore : gitignore 추가

1seyoung committed:
chore : gitignore 변경

1seyoung committed:
chore : gitignore 변경

1seyoung committed:
fix : 시큐리티 제거 및 최소한의 jwt 필터 적용 확인

1seyoung committed:
feat : 교수자 수업 생성 기능 개발 / 학생 등록 기능 개발

1seyoung committed:
refactor: Nginx 설정 추가 및 API 라우팅 개선

1seyoung committed:
chore : 불필요한 프론트 파일 제거

1seyoung committed:
feat : 리뷰 반영 및 오류 수정
- 수업 정보 불러오는 로직 수정
- DB 쿼리 개선

blue000927 commented on src/main/java/com/gatheria/GatheriaApplication.java:
```diff
@@ -2,8 +2,9 @@
+import org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration;
 
-@SpringBootApplication
+@SpringBootApplication(exclude = SecurityAutoConfiguration.class)
```
`exclude = ~` 이 코드 아예 없어도 되지 않나요~?

1seyoung replied:
넵 테스트 중에 작성했던 것 같습니다. 제거하겠습니다~

blue000927 commented on src/main/java/com/gatheria/common/config/PasswordConfig.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
+public class PasswordConfig {
+
+  @Bean
+  public PasswordEncoder passwordEncoder() {
+    return new BCryptPasswordEncoder();
+  }
+}
```
요거도 시큐리티 코드라서 제거 가능해보여요

1seyoung replied:
Spring Security를 사용하지 않는다면, 비밀번호 암호화 및 검증을 처리할 때 어떤 대안을 사용할 수 있을까요?BCryptPasswordEncoder를 지우고 password util 클래스를 만드는게 좋을까요?

blue000927 commented on src/main/java/com/gatheria/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,97 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtAuthenticationFilter implements Filter {
```
고치라는 이야기는 아니고
필터와 인터셉터의 차이에 대해 설명해 주시고, 왜 이 프로젝트에서 필터를 사용하셨는지 설명해주세요 :)

blue000927 commented on src/main/java/com/gatheria/controller/CommonLectureController.java:
```diff
@@ -0,0 +1,36 @@
+  @GetMapping("/{lectureId}")
+  public ResponseEntity<LectureResponseDto> showLecture(@PathVariable Long lectureId,
+      @RequestParam String code,
+      @RequestAttribute("authInfo") AuthInfo authInfo) {
```
여건이 된다면 아규먼트 리졸버를 고려해 봐도 좋아 보입니다.
아규먼트 리졸버를 쓰시면 @Auth AuthInfo authInfo 이런 식으로 어노테이션으로 곧바로 대응이 됩니당

blue000927 commented on src/main/java/com/gatheria/dto/request/LectureCreateRequestDto.java:
```diff
@@ -0,0 +1,12 @@
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+public class LectureCreateRequestDto {
+    private String name;
+}
```
어떤건 RequestDto, 어떤건 Request, 어떤건 Response, 어떤건 ResponseDto인데 통일하면 어떨까요?
request/response or requestDto/responseDto

1seyoung replied:
해당 부분은 팀 생성 관련 코드까지 완료 후에 컨벤션 맞추겠습니다~

blue000927 commented on src/main/java/com/gatheria/service/LectureService.java:
```diff
@@ -0,0 +1,97 @@
+      throw new RuntimeException();
+    }
+    if (authInfo.getRole() == MemberRole.INSTRUCTOR
+        && !Objects.equals(lecture.getInstructorId(), authInfo.getMemberId())) {
```
우리가 공부한 캡슐화를 적용해볼 기회인 것 같습니다.
!Objects.equals(lecture.getInstructorId(), authInfo.getMemberId()) 이 코드 어떻게 수정해 보면 좋을까요?

blue000927 commented on src/main/java/com/gatheria/service/LectureService.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    if (lectureMapper.existEnrollmentByStudentIdAndLectureID(student.getId(), lecture.getId())) {
+      throw new RuntimeException();
```
예외 핸들링 이슈 하나 생성해주세요!
지금은 예외가 모두 RuntimeException이고 예외 메시지도 없으니 나중에 예외가 발생하면 트래킹하기 어려울거예요.

1seyoung replied:
이슈 생성했습니다~!

1seyoung committed:
chore : .gitignore 수정 (db/mysql/data 제외)

1seyoung committed:
chore : PR 리뷰 반영 / 성능 테스트 추가
- 성능 테스트를 위한 도커 파일 수정
- 디비 구조 변경
- AuthInfo 어노테이션 / 리졸버
- 캡슐화 문제 해결
- 스프링 시큐리티 코드 전체 제거

1seyoung committed:
feat : 수업 생성에 제한 인원 추가
